cli/src/main.rs

```diff
@@ -37,172 +54,393 @@ struct Args {
+        let (_, watch_result) = tokio::join!(run_preview_server(), watch_dir());
+
+        watch_result?; // Propagate errors from the watcher and compiler
+
+        return Ok(());
+    }
+
+    // using the watch flag but with some other output format
     if args.watch {
-        watch(&args, &target)?;
-    } else {
-        match compile_file(&args) {
-            Ok((tree, state)) => {
-                let mut stdout = stdout();
-
-                stdout.execute(terminal::Clear(terminal::ClearType::All))?;
-                stdout.execute(cursor::MoveTo(0, 0))?;
-                stdout.execute(style::PrintStyledContent(
-                    "File successfully compiled!\n\n".green(),
-                ))?;
-
-                if !state.warnings.is_empty() {
-                    stdout.execute(style::PrintStyledContent("Warnings:\n".yellow()))?;
-                    for warning in state.warnings {
-                        stdout.execute(style::PrintStyledContent(format!("{warning}").yellow()))?;
-                    }
-                }
+        // Just start the file watcher, but without html live preview
+        // which means that we wil provide None instead of the document and connections
+        return watch_files(None, None, &args, &current_path).await;
+    }
 
-                if !state.errors.is_empty() {
-                    stdout.execute(style::PrintStyledContent("Errors:\n".red()))?;
-                    for error in state.errors {
-                        stdout.execute(style::PrintStyledContent(format!("{error}").red()))?;
-                    }
-                }
+    // Otherwise, if they are using the watcher or live preview
+    // just compile the file once and save it.
+    print_compiling_message()?;
+    let compilation_result = compile_file(&args.input, &args.get_output_format()?);
+    save_result(&compilation_result, &args)?;
+    print_result(&compilation_result, &args)?;
+
+    Ok(())
+}
+
+/// Choose a free port for hosting the html live preview
+fn get_port() -> Result<u16, CliError> {
+    PREVIEW_PORT
+        .get_or_init(|| portpicker::pick_unused_port())
+        .ok_or(CliError::NoFreePorts)
+}
+
+fn get_server_config(
+    document: PreviewDoc,
+    connections: PreviewConnections,
+    current_path: PathBuf,
+) -> impl Filter<Extract = (impl Reply,), Error = Rejection> + Clone + 'static {
+    let connections = warp::any().map(move || connections.clone());
+    let document = warp::any().map(move || document.clone());
+
+    let websocket = warp::path("ws").and(warp::ws()).and(connections).map(
+        |ws: warp::ws::Ws, connections: PreviewConnections| {
+            ws.on_upgrade(move |socket| handle_connection(socket, connections.clone()))
+        },
+    );
+
+    let preview = warp::path::end().and(document).map(|document: PreviewDoc| {
+        // Inject a JS script after the end of the </body> tg
+        let html = document.lock().unwrap();
+        let modified_html = RE_INJECTION.replace(
```

Comment:
This works fine for the special case with HTML, but previewing a latex document won't inject any update code and thus won't do live updates, even though the terminal output says it does.

Reply:
Also, I am unsure why you are using regex in this case. While it might be nice for more advanced matches further down the line, you are replacing "</body>" captured by a regex with the literal "</body>" concatenated with your injection code. Since the second part is still a literal, I think we can do the search-and-replace using a literal as well, that would probably be more performant and won't require regex. As an alternative, consider replacing with the capture group 0 concatenated with your injection.